### PR TITLE
create attendance workflows (DPI) with InMemoryRepo and ZoomAttendeeRepo implementations

### PR DESCRIPTION
**Pull request by sangeethankumar**
Creates an **AttendanceWorkflows** class that contains methods calling an abstract **AttendeeRepo** interface, which has a **ZoomAttendeeRepo** implementation. Have automated tests in place for each workflow method and each **ZoomAttendeeRepo** method, that doesn't call the Zoom API (i.e. goal is to have fast tests)

### CODE REVIEW

File: course_attendance_service/course_attendance_service/workflows/zoom_attendance_repo.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    def __init__(self, zoom_api: ZoomAttendanceApi) -> None:
+        self.api = zoom_api
+        self.access_token = 'ACCESS TOKEN'
+
+    def list_all_attendees(self,workshop_id:str) -> List[str]:
+        report = self.api.get_zoom_participant_report(access_token=self.access_token,meeting_id=workshop_id)
+        participants = report['participants']
+        attendees = []
+        max_duration = np.nanmax([participant['duration'] for participant in participants if participant['status'] == 'in_meeting'])
+        attendance = False
```
Comment (nickdelgrosso) [inline course_attendance_service/course_attendance_service/workflows/zoom_attendance_repo.py:22]: Deciding how a participant should be judged as "attended" or "not attended" for a given session is a higher-level policy.  Would it be possible to move this logic up to the Workflow level?

File: course_attendance_service/course_attendance_service/workflows/test_attendance_workflows.py
Comment (nickdelgrosso) [inline course_attendance_service/course_attendance_service/workflows/test_attendance_workflows.py:1]: Great test approach, looks good!

File: course_attendance_service/course_attendance_service/workflows/zoom_attendance_api.py
Comment (nickdelgrosso) [inline course_attendance_service/course_attendance_service/workflows/zoom_attendance_api.py:1]: Great that ZoomAttendanceAPI is a minimal interface, looks good!

File: DIP.py
Comment (nickdelgrosso) [inline DIP.py:1]: I think this was from our discussion, it's probably best to delete this code.

File: DIP.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+from abc import ABC, abstractmethod
+from typing import NamedTuple, List, Any, Dict
+from pytest import mark
+
+#### TESTS
+
```
Comment (nickdelgrosso) [inline DIP.py:6]: Since there's already `test_attendance_worflows.py`, Maybe it would be good to move this test to that file.

File: DIP.py
Comment (nickdelgrosso) [inline DIP.py:1]: Looks good, good use of the architecture!

**Review by nickdelgrosso — CHANGES_REQUESTED** [review]
Looks good, thank you for this contribution!  I've put some suggestions in the review for changes, but overall I think it looks good and fits the testing needs that the project has.

File: course_attendance_service/course_attendance_service/workflows/zoom_attendance_repo.py (hunk 1 above)
Comment (sangeethankumar) [reply course_attendance_service/course_attendance_service/workflows/zoom_attendance_repo.py]: - Moved logic to Workflow
- Modified Attendee class to accept duration
- attendance and duration are set to None by default and can be set dynamically

**Comment by nickdelgrosso** [comment]
Looks great, thank you @sangeethankumar !

**Pull request merged**